### PR TITLE
Inline source maps

### PR DESCRIPTION
**Pull request by LucasSloan**
We currently handle source maps that are in separate files from the compiled js tsc outputs, but we don't handle source maps in a comment in the compiled js.  This PR handles extracting the source map from the comment, composing it, and adding a new comment with the composed map.

### CODE REVIEW

File: src/tsickle_compiler_host.ts [hunk 1]
```diff
@@ -209,6 +214,15 @@ export class TsickleCompilerHost implements ts.CompilerHost {
     return tscSourceMapGenerator.toString();
   }
 
+  combineInlineSourceMaps(filePath: string, tscOutputText: string): string {
```
Comment (evmar) [inline src/tsickle_compiler_host.ts]: Can you add some jsdoc to this about the parameters?  (e.g. what is tscOutputText)

File: src/tsickle_compiler_host.ts [hunk 2]
```diff
@@ -209,6 +214,15 @@ export class TsickleCompilerHost implements ts.CompilerHost {
...
+    const sourceMapJson = extractInlineSourceMap(tscOutputText);
+    const composedSourceMap = this.combineSourceMaps(filePath, sourceMapJson);
+    const encodedSourceMap = encode(composedSourceMap);
+    return tscOutputText.replace(
+        new RegExp('//# sourceMappingURL=data:application/json;base64,(.*)'),
```
Comment (evmar) [inline src/tsickle_compiler_host.ts]: Should these regexps be more specific?  For example, should they be anchored to the beginning of the line?  Does "." match newlines?

File: src/util.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import {decode} from 'base-64';
 import * as ts from 'typescript';
```
Comment (evmar) [inline src/util.ts]: Somehow these are ending up above the license header, can you put them below?

Comment (LucasSloan) [inline src/util.ts]: Done.

File: src/tsickle_compiler_host.ts (hunk 2 above)
Comment (LucasSloan) [inline src/tsickle_compiler_host.ts]: I've anchored them to the beginning and end of line.

File: src/tsickle_compiler_host.ts (hunk 1 above)
Comment (LucasSloan) [inline src/tsickle_compiler_host.ts]: I've changed tscOutputText to compiledJsWithInlineSourceMap.

File: src/util.ts [hunk 4]
```diff
@@ -91,7 +91,7 @@ export function createOutputRetainingCompilerHost(
 }
 
 export function extractInlineSourceMap(source: string): string {
-  const regex = new RegExp('//# sourceMappingURL=data:application/json;base64,(.*)');
+  const regex = new RegExp('^//# sourceMappingURL=data:application/json;base64,(.*)$', 'm');
   const result = regex.exec(source)!;
```
Comment (evmar) [inline src/util.ts]: What happens if this fails to match?  (Is that possible?)

File: package.json [hunk 5]
```diff
@@ -28,6 +29,7 @@
     "@types/node": "^6.0.38",
     "@types/source-map": "^0.1.27",
     "@types/source-map-support": "^0.2.27",
+    "base-64": "^0.1.0",
```
Comment (evmar) [inline package.json]: Is this library in g3?  We'll have to import it if so.
It might be possible to just use builtin node APIs for base64:
https://nodejs.org/docs/latest/api/buffer.html#buffer_buffers_and_character_encodings

File: test/e2e_source_map_test.ts [hunk 6]
```diff
@@ -160,6 +160,32 @@ describe('source maps', () => {
...
+    sources.set('input.ts', `
+      class X { field: number; }
+      let x : string = 'a string';
+      let y : string = 'another string';
+      let z : string = x + y;`);
+
+    // Run tsickle+TSC to convert inputs to Closure JS files.
+    const {compiledJS, sourceMap} = compile(sources, undefined, undefined, true);
+
+    const {line: stringXLine, column: stringXColumn} = getLineAndColumn(compiledJS, 'a string');
```
Comment (evmar) [inline test/e2e_source_map_test.ts]: I think this would read better as

```
{
  const {line, column} =  ...
  expect ... sourceMap.originalPositionFor({line, column})...
}
{
  const {line, column} =  ...
  expect ... sourceMap.originalPositionFor({line, column})...
}
```

File: src/util.ts (hunk 4 above)
Comment (LucasSloan) [inline src/util.ts]: Tsc should always produce the same source map comment, so since we're protected by a check that the proper flag was passed we should be good.

File: package.json (hunk 5 above)
Comment (LucasSloan) [inline package.json]: That buffer library worked - good catch!

File: test/e2e_source_map_test.ts (hunk 6 above)
Comment (LucasSloan) [inline test/e2e_source_map_test.ts]: That pattern is used a lot of places - mind if I do another PR for that?

File: package.json [hunk 7]
```diff
@@ -18,6 +18,7 @@
     "typescript": "^2.1.0"
   },
   "devDependencies": {
+    "@types/base-64": "^0.1.0",
```
Comment (evmar) [inline package.json]: drop this

File: test/e2e_source_map_test.ts [hunk 8]
```diff
@@ -229,21 +255,32 @@ function tsickleCompiler(
...
 
+  let compilerOptions: ts.CompilerOptions;
+  if (inlineSourceMap) {
+    compilerOptions = {
+      inlineSourceMap: inlineSourceMap,
+      outFile: outFile,
+      experimentalDecorators: true
+    };
+  } else {
+    compilerOptions = {sourceMap: true, outFile: outFile, experimentalDecorators: true};
```
Comment (evmar) [inline test/e2e_source_map_test.ts]: if you write this with an extra comma:
```
experimentalDecorators: true,};
```
clang-format will lay it out linearly which makes it match the line above.
